source/dhtslib/threadpool.d

```diff
@@ -0,0 +1,482 @@
+/**
+This module provides multithreading support through htslib. It allows for mapping delagates,
+and functions via a range interface with `parallelMap`. Crucially, the htslib task pool performs
+all work in order. All functions and interfaces below should as well.
+*/
+module dhtslib.threadpool;
+import dhtslib.memory;
+import htslib.hts;
+import htslib.thread_pool;
+import std.algorithm : map;
+import std.parallelism : totalCPUs;
+import std.traits;
+import std.range;
+import std.meta;
+import core.stdc.stdlib : calloc, free;
+import core.atomic : atomicOp;
+
+/// global thread pool
+__gshared ThreadPool globalPool;
+
+/// ensure global pool is initialized
+/// NOTE: only sets threads if the global pool is not already initialized
+void enforceGlobalThreadPool(int threads = -2) @trusted nothrow @nogc 
+{
+    if(globalPool.tpool && (globalPool.threads == threads || threads == -2)) {
+        return;
+    } else
+        globalPool = ThreadPool(threads, true);   
+}
+
+private template noUnsharedAliasing(T)
+{
+    enum bool noUnsharedAliasing = !hasUnsharedAliasing!T;
+}
+
+// This template tests whether a function may be executed in parallel from
+// @safe code via Task.executeInNewThread().  There is an additional
+// requirement for executing it via a TaskPool.  (See isSafeReturn).
+private template isSafeTask(F)
+{
+    enum bool isSafeTask =
+        (functionAttributes!F & (FunctionAttribute.safe | FunctionAttribute.trusted)) != 0 &&
+        (functionAttributes!F & FunctionAttribute.ref_) == 0 &&
+        (isFunctionPointer!F || !hasUnsharedAliasing!F) &&
+        allSatisfy!(noUnsharedAliasing, Parameters!F);
+}
+
+/// Thread pool instance. Wraps htsThreadPool and hts_tpool. 
+struct ThreadPool 
+{
+    htsThreadPool htsPool;
+    HtsTPool _tpool;
+    int threads;
+    bool isGlobal;
+    bool owned;
+    @nogc @trusted nothrow: 
+    this(int threads, bool global = false) {
+        this.threads = threads;
+        if (threads == -1)
+        {
+            this.threads = totalCPUs;
+        }
+        this._tpool = HtsTPool(hts_tpool_init(cast(int)this.threads));
+        this.htsPool.pool = this._tpool;
+        this.htsPool.qsize = 0;
+        this.isGlobal = global;
+        this.owned = true;
+    }
+
+    this(this) {
+        if(this.owned) this._tpool = _tpool;
+    }
+
+    @property tpool() {
+        return this.htsPool.pool;
+    }
+
+}
+
+
+/**
+Below code is ported from std.parallelism to work with htslib's thread pool.
```

Comment:
I'm explicitly not reviewing any of the code directly adapted from std.parallelism

Reply:
Most of `Task` is ripped directly because I needed their compile time logic to make it work with any function/delegate. I also tried to port `parallel` that could be used in a `foreach` statement, but it was too annoying and I don't really need it anyway.